### PR TITLE
Added support for skia platform

### PR DESCRIPTION
**Pull request by munezbn**
Added support for skia platform

### CODE REVIEW

File: skia/BUILD.gn [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  sources = [
+    "RSkComponentProviderBVLinearGradient.cpp",
+    "RSkComponentProviderBVLinearGradient.h",
+    "RSkComponentBVLinearGradient.cpp",
+    "RSkComponentBVLinearGradient.h",
+    "react/renderer/components/rnx-gradient/ShadowNodes.cpp",
+    "react/renderer/components/rnx-gradient/Props.cpp",
+  ]
+
+  # RNSVG cpp sources
```
Comment (shefayaparwin) [inline skia/BUILD.gn]: change Comment  to RNLinerGradient

File: skia/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                  "//react-native/ReactCommon/yoga/",
+                  "//rns_shell/",
+                  "//ReactSkia/",
+                  ]
+}
+
+rn_cxx_component("BVLinearGradient") {
+
+  # RNS Fabric Component and its dependencies
+  sources = [
```
Comment (shefayaparwin) [inline skia/BUILD.gn:27]: Can we maintain separate sources for  react common & fabric component dependencies?

File: skia/RSkComponentBVLinearGradient.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  RNS_LOG_WARN("[RSkComponentBVLinearGradient][handleCommand] commandName : "<< commandName);
+  RNS_LOG_NOT_IMPL;
+  return;
+}
+
+RnsShell::LayerInvalidateMask  RSkComponentBVLinearGradient::updateComponentProps(SharedProps newViewProps, bool forceUpdate) {
+  RnsShell::LayerInvalidateMask invalidateMask = RnsShell::LayerInvalidateNone;
+
+  auto const &newBVLinearGradientProps = *std::static_pointer_cast<LinearGradientViewProps const>(newViewProps);
+  auto const &oldBVLinearGradientProps = *std::static_pointer_cast<LinearGradientViewProps const>(getComponentData().props);
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: Can we drop extracting Old props, as it is not used.

File: skia/RSkComponentBVLinearGradient.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  Rect frame = component.layoutMetrics.frame;
+  SharedColor bgColour = linearGradientProps.backgroundColor;
+  SkScalar frameX = frame.origin.x;
+  SkScalar frameY = frame.origin.y;
+  SkScalar frameWidth = frame.size.width;
+  SkScalar frameHeight = frame.size.height;
+
+  // If background color is transparent then need to find atleast one non-transparent colour in gradient colours so that we can draw shadow properly.
+  for(int i = 0; (!(bgColour && (colorComponentsFromColor(bgColour).alpha != 0.0)) && i < numColors); i++) {
+    bgColour = linearGradientProps.colors.at(i);
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp:71]: Is this condition needs to break on finding single color value ?

File: skia/react/renderer/components/rnx-gradient/RCTComponentViewHelpers.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/**
```
Comment (shefayaparwin) [inline skia/react/renderer/components/rnx-gradient/RCTComponentViewHelpers.h:1]: Is this File needed for RN LinearGradient

File: skia/react/renderer/components/rnx-gradient/Props.h [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+ */
+#pragma once
+
+#include <jsi/jsi.h>
+#include <react/renderer/components/view/ViewProps.h>
+#include <react/renderer/core/propsConversions.h>
+#include <react/renderer/graphics/Color.h>
+#include <vector>
+
+#include "ReactSkia/views/common/RSkConversion.h"
```
Comment (shefayaparwin) [inline skia/react/renderer/components/rnx-gradient/Props.h:18]: Is this header file Needed ?

File: skia/RSkComponentBVLinearGradient.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+namespace react {
+
+using namespace std;
+
+class RSkComponentBVLinearGradient : public RSkComponent {
+ public:
+  RSkComponentBVLinearGradient(const ShadowView &shadowView);
+  ~RSkComponentBVLinearGradient() = default;
+  RnsShell::LayerInvalidateMask updateComponentProps(SharedProps newviewProps,bool forceUpdate) override;
+  void handleCommand(std::string commandName, folly::dynamic args) override;
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.h]: If this Module doesn't have to process any commands, can this override of handleCommand be avoided. ?

File: skia/RSkComponentBVLinearGradient.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    if(locationsMapAvailable) {
+      pos[i] = linearGradientProps.locations.at(i);
+    }
+  }
+
+  SkPoint startEndPoints[] = { //Initialize with default value
+    {(0.5 * frameWidth) + frameX, (0 * frameHeight) + frameY},
+    {(0.5 * frameWidth)+ frameX, (1 * frameHeight) + frameY}};
+
+  SkMatrix matrx = SkMatrix::I();
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp:102]: looks like a type error on variable name [matrx-->matrix]

File: skia/RSkComponentBVLinearGradient.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  auto lgs = SkGradientShader::MakeLinear(startEndPoints,
+                                          colors,
+                                          (locationsMapAvailable ? pos : nullptr),
+                                          numColors,
+                                          SkTileMode::kClamp,
+                                          0/*flags*/, &matrx);
+
+  SkPaint paint;
+  paint.setShader(lgs);
+  canvas->drawRoundRect({frameX, frameY, frameX+frameWidth, frameY+frameHeight},
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: If View can take radii on all four Sides, Then drawRRect would be the Right Choice.

File: skia/RSkComponentBVLinearGradient.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                layerRef->opacity,
+                layerRef->shadowImageFilter,layerRef->shadowMaskFilter);
+  }
+
+  RNS_LOG_DEBUG(std::this_thread::get_id() << "] Layer " << layer().get() <<" Draw Picture Here with Frame  [" << frameX << "," << frameY << "," << frameWidth << "," << frameHeight << "]");
+
+  SkColor colors[numColors];
+  SkScalar pos[linearGradientProps.locations.size()];
+
+  bool locationsMapAvailable = (linearGradientProps.locations.size() == numColors);
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp:75]: will APP allows to specify mismatched  position & color info , if then Instead of ignoring position info, can we use the Matched Info & ignore others.
Or Is Skia capable of handling  mismatch in color & position by itself

File: skia/RSkComponentBVLinearGradient.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                bgColour, // Used only to decide on what type of shadow to draw.
+                layerRef->shadowColor,layerRef->shadowOffset,layerRef->shadowOpacity,
+                layerRef->opacity,
+                layerRef->shadowImageFilter,layerRef->shadowMaskFilter);
+  }
+
+  RNS_LOG_DEBUG(std::this_thread::get_id() << "] Layer " << layer().get() <<" Draw Picture Here with Frame  [" << frameX << "," << frameY << "," << frameWidth << "," << frameHeight << "]");
+
+  SkColor colors[numColors];
+  SkScalar pos[linearGradientProps.locations.size()];
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp:73]: Is it possible App allows to specify only Single colour against the expectation, If then can the shader creation can be avoided under colorSize > 1 condition ?

File: skia/RSkComponentBVLinearGradient.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  bool locationsMapAvailable = (linearGradientProps.locations.size() == numColors);
+
+  for (std::size_t i = 0; i != numColors; ++i) {
+    colors[i] = RSkColorFromSharedColor(linearGradientProps.colors.at(i), SK_ColorTRANSPARENT);
+    if(locationsMapAvailable) {
+      pos[i] = linearGradientProps.locations.at(i);
+    }
+  }
+
+  SkPoint startEndPoints[] = { //Initialize with default value
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: Can we specify the type of default gradient Like : Horizontal or Vertical Gradient ...in comment

File: skia/RSkComponentProviderBVLinearGradient.cpp [hunk 13]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (shefayaparwin) [inline skia/RSkComponentProviderBVLinearGradient.cpp:1]: Can we change this provider to macro based . Assuming the changes are made in ReactNativeSkia branch
as RNS_SETUP_COMPOMNENT_PROVIDER

File: skia/RSkComponentBVLinearGradient.cpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+ * Copyright (C) 1994-2023 OpenTV, Inc. and Nagravision S.A.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "include/core/SkPaint.h"
+#include "include/effects/SkGradientShader.h"
+
+#include "cxxreact/ModuleRegistry.h"
```
Comment (chatrapathi06) [inline skia/RSkComponentBVLinearGradient.cpp]: Is ModuleRegistry.h used any where, if not we can remove

File: skia/RSkComponentBVLinearGradient.cpp [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+#include "react/renderer/components/rnx-gradient/Props.h"
+#include "rns_shell/compositor/layers/PictureLayer.h"
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: IS this Header Needed ?

File: skia/RSkComponentBVLinearGradient.cpp (hunk 14 above)
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: Is This Header Needed in this File ?

File: skia/RSkComponentBVLinearGradient.cpp [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+  return;
+}
+
+RnsShell::LayerInvalidateMask  RSkComponentBVLinearGradient::updateComponentProps(SharedProps newViewProps, bool forceUpdate) {
+  RnsShell::LayerInvalidateMask invalidateMask = RnsShell::LayerInvalidateNone;
+
+  auto const &newBVLinearGradientProps = *std::static_pointer_cast<LinearGradientViewProps const>(newViewProps);
+  auto const &oldBVLinearGradientProps = *std::static_pointer_cast<LinearGradientViewProps const>(getComponentData().props);
+
+  RNS_LOG_INFO(" Start: [" << newBVLinearGradientProps.start.x << "," << newBVLinearGradientProps.start.y <<
```
Comment (chatrapathi06) [inline skia/RSkComponentBVLinearGradient.cpp]: can we use  RNS_LOG_DEBUG  instead of RNS_LOG_INFO.

File: skia/RSkComponentBVLinearGradient.cpp [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+#include "include/core/SkPaint.h"
+#include "include/effects/SkGradientShader.h"
+
+#include "cxxreact/ModuleRegistry.h"
+
+#include "react/renderer/components/rnx-gradient/Props.h"
+#include "rns_shell/compositor/layers/PictureLayer.h"
+
+#include "ReactSkia/views/common/RSkDrawUtils.h"
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: Looks like APi from RSKDrawUtils is not used. So is THis needed Here ?

File: skia/RSkComponentBVLinearGradient.cpp [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+
+#include "RSkComponentBVLinearGradient.h"
+
+namespace facebook {
+namespace react {
+
+RSkComponentBVLinearGradient::RSkComponentBVLinearGradient(const ShadowView &shadowView)
+    : RSkComponent(shadowView) {
+  RNS_LOG_TODO("Listen to notifications about module invalidation default notification center : RCTWillInvalidateModulesNotification");
```
Comment (shefayaparwin) [inline skia/RSkComponentBVLinearGradient.cpp]: Just Need to check on it . Do We Need this ?

File: skia/RSkComponentBVLinearGradient.h [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+#include "ReactSkia/components/RSkComponent.h"
+#include "ReactSkia/utils/RnsUtils.h"
+
+// When useAngle is true convert it to cartesian degrees
+#define RNS_BVLG_CARTESIAN_ANGLE_CORRECTION 180
+
+namespace facebook {
+namespace react {
+
+using namespace std;
```
Comment (chatrapathi06) [inline skia/RSkComponentBVLinearGradient.h]: is using namespace std; required ? since all the placing we are using std::.

File: skia/RSkComponentBVLinearGradient.cpp (hunk 14 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Removed.

File: skia/BUILD.gn (hunk 1 above)
Comment (munezbn) [reply skia/BUILD.gn]: Removed this empty block

File: skia/RSkComponentBVLinearGradient.cpp (hunk 14 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: removed

File: skia/RSkComponentBVLinearGradient.cpp (hunk 17 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Removed

File: skia/RSkComponentBVLinearGradient.cpp (hunk 15 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Removed

File: skia/RSkComponentBVLinearGradient.cpp (hunk 18 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Not required for this component. Removed.

File: skia/RSkComponentBVLinearGradient.cpp (hunk 3 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: done

File: skia/RSkComponentBVLinearGradient.cpp (hunk 16 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Done

File: skia/RSkComponentBVLinearGradient.cpp (hunk 11 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp:73]: As per component minimum 2 colours must be provided. Even if we don't pass the SkGradient will take care of that

File: skia/RSkComponentBVLinearGradient.cpp (hunk 10 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp:75]: As per component it is mandatory to provide location for each colour. Hence we are ignoring the locations and let skia to use default locations. Infact even android ignores if colour size and location size doesn't match

File: skia/RSkComponentBVLinearGradient.cpp (hunk 12 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Gradient usually doesn't speak about horizontal or vertical etc. It talks only in term,s of start-end points or angle .
If we consider horizontal/Vertical then in that also we have left to right or right to left or top to bottom or bottom to top..
I have updated comment

File: skia/RSkComponentBVLinearGradient.h (hunk 7 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.h]: Removed

File: skia/react/renderer/components/rnx-gradient/Props.h (hunk 6 above)
Comment (munezbn) [reply skia/react/renderer/components/rnx-gradient/Props.h:18]: Yes we are using Colour related API from this file

File: skia/react/renderer/components/rnx-gradient/RCTComponentViewHelpers.h (hunk 5 above)
Comment (munezbn) [reply skia/react/renderer/components/rnx-gradient/RCTComponentViewHelpers.h:1]: No, But his code is generated by codegen hence not touching it

File: skia/RSkComponentBVLinearGradient.h (hunk 19 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.h]: Removed

File: skia/RSkComponentBVLinearGradient.cpp (hunk 9 above)
Comment (munezbn) [reply skia/RSkComponentBVLinearGradient.cpp]: Done

**Review by shefayaparwin — APPROVED** [review]

**Review by chatrapathi06 — APPROVED** [review]
